File: sentimental-credit/credit.py
from cs50 import get_string

def main():
    # Prompt user for credit card number
    card_number = get_string("Number: ")

    # Perform Luhn's Algorithm
    checksum = calculate_checksum(card_number)
    if checksum % 10 != 0:
        print("INVALID")
        return

    # Check card type
    card_length = len(card_number)
    if card_length == 15 and card_number[0] == "3" and (card_number[1] == "4" or card_number[1] == "7"):
        print("AMEX")
    elif card_length == 16 and card_number[0] == "5" and int(card_number[1]) in range(1, 6):
        print("MASTERCARD")
    elif (card_length == 13 or card_length == 16) and card_number[0] == "4":
        print("VISA")
    else:
        print("INVALID")

def calculate_checksum(card_number):
    total = 0
    for i in range(len(card_number)-2, -1, -2):
        digit = int(card_number[i]) * 2
        if digit > 9:
            digit = digit // 10 + digit % 10
        total += digit
    for i in range(len(card_number)-1, -1, -2):
        total += int(card_number[i])
    return total

if __name__ == "__main__":
    main()

File: credit/credit.c
#include <stdio.h>
#include <cs50.h>

int get_digit_sum(int number);
bool is_valid_card(long card_number);
string get_card_type(long card_number);

int main(void)
{
    long card_number = get_long("Number: ");

    if (is_valid_card(card_number))
    {
        string card_type = get_card_type(card_number);
        printf("%s\n", card_type);
    }
    else
    {
        printf("INVALID\n");
    }
}

int get_digit_sum(int number)
{
    int sum = 0;

    while (number > 0)
    {
        sum += number % 10;
        number /= 10;
    }

    return sum;
}

bool is_valid_card(long card_number)
{
    int sum = 0;
    bool multiply = false;

    while (card_number > 0)
    {
        int digit = card_number % 10;

        if (multiply)
        {
            digit *= 2;
            sum += get_digit_sum(digit);
        }
        else
        {
            sum += digit;
        }

        card_number /= 10;
        multiply = !multiply;
    }

    return sum % 10 == 0;
}

string get_card_type(long card_number)
{
    int first_two_digits = card_number / 10000000000000;
    int first_digit = card_number / 1000000000000000;

    if ((first_two_digits == 34 || first_two_digits == 37) && (first_digit == 0))
    {
        return "AMEX";
    }
    else if (first_two_digits >= 51 && first_two_digits <= 55)
    {
        return "MASTERCARD";
    }
    else if ((first_digit == 4) && (first_two_digits != 49 && first_two_digits != 48 && first_two_digits != 47 && first_two_digits != 46 && first_two_digits != 45))
    {
        return "VISA";
    }
    else
    {
        return "INVALID";
    }
}

